### PR TITLE
Rewrites closure detection to work without using regular expressions

### PR DESCRIPTION
**Pull request by bkchr**
- Remove all regex occurrences
- Use the syntex crate to parse the actual closure parameters
- Add test for issue 754

Fixes: #754

### CODE REVIEW

File: src/racer/util.rs [hunk 1]
```diff
@@ -70,6 +70,27 @@ pub fn symbol_matches(stype: SearchType, searchstr: &str, candidate: &str) -> bo
     }
 }
 
+// Try to valid if the given scope contains a valid closure arg scope.
```
Comment (TedDriggs) [inline src/racer/util.rs]: NIT: Please use `///` so this gets picked up as a doc comment.

File: src/racer/util.rs [hunk 2]
```diff
@@ -70,6 +70,27 @@ pub fn symbol_matches(stype: SearchType, searchstr: &str, candidate: &str) -> bo
...
+    let right_pipe = if let Some(pos) = rest_scope.find('|') {
+        left_pipe + 1 + pos
+    } else {
+        return None;
+    };
+
+    let pipe_scope = &scope_src[left_pipe..right_pipe+1];
+
+    // If we found not forbidden characters --> no closure definition
+    let forbidden_chars = [ ";", "{", "}" ];
```
Comment (TedDriggs) [inline src/racer/util.rs]: Struct patterns are allowed in closure arguments; **[here's](https://play.rust-lang.org/?gist=c5ad305c6c473028507a1926ff225664&version=stable)** an example. Please either a) have that work, and add a test to demonstrate it, or b) add a note to this method's doc comment calling out that struct patterns in closure arguments aren't supported.

File: tests/system.rs [hunk 3]
```diff
@@ -3330,6 +3330,34 @@ fn closure_bracket_scope_nested_match_outside() {
 }
 
 #[test]
+fn issue_754() {
```
Comment (TedDriggs) [inline tests/system.rs]: Please make the test name descriptive of what it demonstrates. Adding a link to the GitHub issue in the test's doc comment would be great though 😃

File: src/racer/nameres.rs [hunk 4]
```diff
@@ -946,68 +946,49 @@ pub fn search_scope(start: Point, point: Point, src: Src,
...
-                local: true,
-                mtype: FnArg,
-                contextstr: cap.get(0).unwrap().as_str().to_owned(),
-                generic_args: Vec::new(),
-                generic_types: Vec::new(),
-                docs: String::new(),
-            })
-        } else {
-            None
+    if let Some((left_pipe, _, pipe_scope)) = closure_valid_arg_scope(scope_src) {
```
Comment (TedDriggs) [inline src/racer/nameres.rs]: Can you please add a `trace!` statement here which reports that the function has found a closure?

Comment (bkchr) [inline src/racer/nameres.rs]: Added :)

**Comment by bkchr** [comment]
Addressed your requests. Especially thanks for finding that curly bracket use case in a closure argument scope :D

**Comment by TedDriggs** [comment]
Thanks for adding the tests! I'd like @jwilm to look at it, but otherwise LGTM

**Comment by bkchr** [comment]
Ping @jwilm ^^

**Comment by bkchr** [comment]
Fixed the merge conflict! So it should be ready to merge! :)

**Pull request merged**